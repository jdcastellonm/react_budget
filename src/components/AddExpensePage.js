import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {startAddExpense} from '../actions/expenses';

// refactor to class to use mapDispatchToProps and make testing easier
// 'expense' variable comes from the ExpenseForm
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Save Expense</h1>
                    </div>
                </div>
                <div className="content-container"><ExpenseForm onSubmit={this.onSubmit}/></div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({ // allows to call dispatch using simpler props.onSubmit syntax in the class, for easier testing
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
