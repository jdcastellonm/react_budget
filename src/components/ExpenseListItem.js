import React from 'react';
import {Link} from 'react-router-dom';

// the spread can also be destructured like {description, amount, createdAt}
// instead of using props.*
const ExpenseListItem = (props) => (
    <div>
        <Link to={`/edit/${props.id}`}>
            <h3>{props.description}</h3>
        </Link>
        <p>{props.amount} -- {props.createdAt}</p>
    </div>
)
// export default connect()(ExpenseListItem); // access to state directly is not needed here, just the dispatch() (used when the remove button was here)
export default ExpenseListItem;