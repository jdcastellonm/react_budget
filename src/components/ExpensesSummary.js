import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectedExpenses from '../selectors/expenses';
import getExpenseTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => { // destructuring props
    const word = expenseCount === 1 ? 'expense' : 'expenses';
    const totalFormatted = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
            <h1>Viewing {expenseCount} {word} totalling {totalFormatted}</h1>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectedExpenses(state.expenses, state.filters); // reminder this returns a filtered array
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpenseTotal(visibleExpenses) // get the sum from the filtered array
    }
};

export default connect(mapStateToProps)(ExpensesSummary);