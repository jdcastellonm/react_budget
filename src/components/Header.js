import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Budget</h1>
        <NavLink exact={true} activeClassName='is-active' to='/'>Home</NavLink>
        <NavLink exact={true} activeClassName='is-active' to='/create'>Add Expense</NavLink>
        <NavLink exact={true} activeClassName='is-active' to='/edit'>Edit Expense</NavLink>
        <NavLink exact={true} activeClassName='is-active' to='/help'>Help</NavLink>
    </header>
)
export default Header;