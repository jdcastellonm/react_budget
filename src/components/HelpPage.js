import React from 'react';

const HelpPage = () => (
    <div>
        help component
    </div>
)
export default HelpPage;