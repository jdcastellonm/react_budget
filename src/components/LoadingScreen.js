import React from 'react';

const loadingScreen = () => (
    <div className="loader">
        <img className="loader__image" src="/images/loader.gif"/>
    </div>
);

export default loadingScreen;
