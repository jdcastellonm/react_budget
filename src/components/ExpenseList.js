import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (<p>No expenses.</p>) :
            (props.expenses.map((current) => {
                return <ExpenseListItem key={current.id} {...current}/>
            }))
        }
    </div>
);


// this is a hoc/decorator. the wrapper function gets returned from connect(), which is then immediately run to wrap
// the ExpenseList component. likely written this way to be able to make use of the state when preparing the wrapper
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters) // calling the sort by filters function here before mapping to props
    }
};
export default connect(mapStateToProps)(ExpenseList);