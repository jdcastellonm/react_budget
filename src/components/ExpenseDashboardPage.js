import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary'; // the default export is connected to the store so no need to pass in props

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
)
export default ExpenseDashboardPage;