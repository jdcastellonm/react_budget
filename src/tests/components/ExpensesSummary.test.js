import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('render summary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={234}/>);
    expect(wrapper).toMatchSnapshot();
});
test('render summary with expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={21} expensesTotal={23489234}/>);
    expect(wrapper).toMatchSnapshot();
});