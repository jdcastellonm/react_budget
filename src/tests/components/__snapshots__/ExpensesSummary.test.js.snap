// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render summary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.34
  </h1>
</div>
`;

exports[`render summary with expenses 1`] = `
<div>
  <h1>
    Viewing 
    21
     
    expenses
     totalling 
    $234,892.34
  </h1>
</div>
`;
